### ISSUE TITLE
Add ability to set the shell 

### ISSUE THREAD

**Issue by scottTomaszewski** [jetify-com/devbox#2148]
### What problem are you trying to solve?

In the `init_hook` I would like to source some scripts containing helper utilities that are written in bash.  In `devbox run` the shell is `sh` whereas `devbox shell` the shell is inherited from the caller's shell.  

This behavior was confirmed in discord:
```
We designed it this way initially to make run calls more portable  and consistent. The thinking was that devbox run (especially if used in CI) would use a consistent shell and not be affected by a users specific settings, while devbox shell would use the users shell
```
which make sense, but if I wanted a consistent shell environment across both `devbox shell` and `devbox run` it would be nice to have a way to set that.

More discussion:
- https://discord.com/channels/903306922852245526/1234953700217323541
- https://discord.com/channels/903306922852245526/1251154511746039828

### What solution would you like?

Ideally, the solution would be to have support for overriding the default shell for `devbox shell` and `devbox run`.  For (a contrived) example:
```
{
  "$schema":  "https://raw.githubusercontent.com/jetify-com/devbox/0.11.0/.schema/devbox.schema.json",
  "packages": ["bash@latest"],
  "shell": {
    "exec": {
      "shell": "/bin/bash --rcfile something.bash",
      "run": "BASH_ENV=something.bash /bin/bash -c"
    },
    "init_hook": [
      "ps -p $$",
      "source .bootstrap.sh"
    ],
    ...
```

- Splitting out `.shell.exec.shell` and `.shell.exec.run` allows for flexibility, but if even without the split my use case is satisfied
- Allowing arguments (not just a shell path) allows for flags to be passed and env vars to be set if necessary
- I understand that this makes for a bit of a mess with inheriting the existing shell.  Overriding the shell with this may assume that the `--pure` flag is automatically set

### Alternatives you've considered

- Set the `SHELL` in the devbox.json `env` (no effect)
- `chsh` in the `init_hook` (requires password or mucking with global filesystem)
- `eval bash`, `bash -l`, and other means to starting the desired shell (creates a subshell which doesnt run the `devbox run` scripts)
- rewriting all our helpers in sh (if I declare `bash` as a package, I should be able to use it in scripts)
- Prefixing all script lines with `bash` to force the shell (verbose and problematic with the creation of subshells)

**Comment by Lagoja** [comment]
I think this is a really good suggestion, and it would help with some shell incompatibility bugs we've seen in the past. 

I think I would tweak the proposal to allow the user to specify a shell package they want to install and use as well. For example (not final):

```json
  "shell": {
    "package": "bash@5.1"
    "exec": {
      "shell": "bash --rcfile something.bash",
      "run": "BASH_ENV=something.bash bash -c"
    },
```

Would install Bash 5.1 from nixpkgs and then run shell/commands/services in that shell. This way you could write zsh/bash/fish specific scripts, and have it work across systems regardless of the user's host shell.

**Comment by mvanholsteijn** [comment]
It is counter-intuitive to devbox that the host shell is taken, given that the we expect an  ***isolated, reproducible development environments that runs anywhere***

Bash is installed by devbox and is used when running `--pure`. It would be logical to use this as the default shell for all cases. 
An override for the default shell would be optional.

```nix
  "shell": {
    "package": "zsh@5.9"
    "cmd": "bin/zsh -c"
 }
```

Allowing for separate shells for shell, init and scripts can become pretty confusing.

**Comment by scottTomaszewski** [comment]
> Allowing for separate shells for `shell`, `init_hooks` and `run` can become pretty confusing.

I agree, but from what I understand `shell` and `run` execute with different interactive/login settings which may require different `cmd` options.  For example, `--rcfile` vs `BASH_ENV`.  I could be wrong.

**Comment by mvanholsteijn** [comment]
@scottTomaszewski After diving a litter deeper,  I suspect that the need for different configurations for `run` and `shell` 
commands is the init_hook commands. These commands need to be sourced in for both the shell and run commands. 

Should you be able to reference the init_hook command script in the exec specification?

```
  "shell": {
    "package": "bash@5.1"
    "exec": {
      "shell": "bash --rcfile ${init_hook_commands}",
      "run": "BASH_ENV=${init_hook_commands} bash -c"
    },
```

**Comment by svallory** [comment]
I've wasted a lot of time trying to figure out why some things worked if I ran from a devbox shell but did not when using `devbox run` :(

Initially, I expected a completely isolated system (even with different $HOME dir), when I noticed it wasn't, the expectation was that things where running in my shell (zsh).

I tailored all the scripts to zsh, but found out that scripts run on `sh` and `--pure` shells run on bash. This has been  very frustrating journey so far.

Would it really break everything to have a setting on `devbox.json` that specifies the shell to be used for everything?

**Comment by sascha-coenen** [comment]
+1  for adding this feature

having this ability in devbox is without question the highest item on my wishlist. 
Please, please please, can we have this feature in devbox as soon as humanly possible?

**Comment by Nugine** [comment]
I'm new to devbox. My first try makes me confused about which shell devbox is using.

My login shell is `bash` but I use `fish` in gnome terminal and vscode.
`devbox shell` runs `bash` because `SHELL` env variable is `/bin/bash`. It does not inherit the caller shell (`fish`).

I cannot find a way to configure this behavior. And finally I find this issue.

**Comment by scottTomaszewski** [comment]
No, my hacky workaround is to have a `source .utils/.utilsrc` in the `init_hook` and everything delegates to a justfile that has:
```
export BASH_ENV := ".utils/.utilsrc"
set shell := ["bash", "-c"]
```

Im using justfile for everything anyway and devbox is just used for package management... There is probably a better way to do this, but i haven't gotten around to figuring it out

**Comment by svallory** [comment]
Can't believe this is still open. This issue is what made me abandon devbox :/ Whiich is very sad because appart from this, it's a very cool product. And please, devbox team, understand that it is not about being picky. The amount of familiarity I have with ZSH (in my case) and resources I would be able to reuse simply save me more time then dealing with bash to setup a devbox.

And this is not about shell X is superior. It's just that the cost of adopting devbox when I have to learn bash (a shell I don't use) and port a bunch of stuff is simply too high

**Comment by barraponto** [comment]
@svallory yeah, it sort of was the reason i left too.
but OTOH, it kind of changed my path: maybe i don't need to enforce the shell.
maybe all I need is to make the tooling available and leave the user to their own devices.

of course, if a shell script is part of the project, I can make the shell available for that script.
though if i'm providing the tools, i can do better than shell script.

**Comment by ka-rostom** [comment]
Just to follow up (and without any attitude or sass), is there any interest, intent, or activity on jetify team's end to implement this, or nah?

**Comment by svallory** [comment]
> [@svallory](https://github.com/svallory) yeah, it sort of was the reason i left too. but OTOH, it kind of changed my path: maybe i don't need to enforce the shell. maybe all I need is to make the tooling available and leave the user to their own devices.

I actually think the user should have a choice. My case was the opposite of yours. I was forced to do things in bash because that's the shell devbox uses and there's no way around it.

I don't remember exactly what it was, but, for example, you cannot set an environment variable from a subshell. So executing a script in a zsh subshell was not an option and I would have to migrate several scripts to bash, so I simply went to a competitor.

**Comment by apgrucza** [comment]
@svallory I worked around the fact that you cannot set an environment variable from a subshell by running the below command in the init hook:

```shell
eval "$(print_init_env.sh || echo false)"
```

`print_init_env.sh` can be written in any shell language you like (using a shebang), but it must output a script that sets the variables, e.g.:

```
export FOO=foo
export BAR=bar
```

But I agree it would be so much easier if Devbox allowed the init hook shell to be configured.

**Comment by svallory** [comment]
@apgrucza that's a clever trick, but I was modularizing the configuration using devbox plugins. I was getting the values from the config, so it would be a huge hassle and potentially a security vulnerability to run `eval`

**Comment by davidmarble** [comment]
@svallory Mind sharing what you use instead of devbox? I'm a new user and running into the issues you mention. Feel free to email if you prefer not to share on this issue.

**Comment by svallory** [comment]
> [@svallory](https://github.com/svallory) Mind sharing what you use instead of devbox? I'm a new user and running into the issues you mention. Feel free to email if you prefer not to share on this issue.

I don't mind (I just hope this isn't rude)... I'm using [flox.dev](//flox.dev), it supports bash, zsh, fish, and tcsh.

**Comment by barraponto** [comment]
fwiw, i went with flox.dev too. the [layering environments](https://flox.dev/docs/tutorials/layering-multiple-environments/) feature made me realize all I needed was a single layer with the tools -- leave the actual shell to the user.

it does support environment variables too, btw.

**Comment by HiroyukiMakita** [comment]
> [@svallory](https://github.com/svallory) I worked around the fact that you cannot set an environment variable from a subshell by running the below command in the init hook:
> 
> eval "$(print_init_env.sh || echo false)"
> `print_init_env.sh` can be written in any shell language you like (using a shebang), but it must output a script that sets the variables, e.g.:
> 
> ```
> export FOO=foo
> export BAR=bar
> ```
> 
> But I agree it would be so much easier if Devbox allowed the init hook shell to be configured.


Running the following command in the init hook enables the use of environment variables defined in files such as `.env` located in the project root.

This method works with both `devbox shell` and `devbox run`.


```
    "init_hook": [
      "cd $DEVBOX_PROJECT_ROOT",
      "set -a",
      "[ -f .env ] && . \"$DEVBOX_PROJECT_ROOT/.env\"",
      "set +a"
    ]
```


By the way, I'm all for adding the ability to change the shell, and I'm looking forward to it too.

**Comment by nhuray** [comment]
+1 for adding this feature

**Comment by dn-jmp** [comment]
Would love to see this feature, I want consistency in the behavior offered to my colleagues, including the shell behaviors.

**Comment by worldofgeese** [comment]
## Workaround we're using

We've been working around this limitation for months. Here's the hack we use in `devbox.json`:

```json
{
  "shell": {
    "init_hook": [
      "# ... other init commands ...",
      "",
      "# Launch xonsh for interactive devbox shell sessions",
      "# (devbox run scripts stay under sh — this only fires for 'devbox shell')",
      "if [ -t 0 ] && [ -z \"\" ] && command -v xonsh >/dev/null 2>&1; then",
      "  exec xonsh",
      "fi"
    ]
  }
}
```

The pattern works by:

1. **Detecting interactive TTY**: `[ -t 0 ]` checks if stdin is a terminal
2. **Skipping `devbox run`**: We check `[ -z \"\"\" \` to avoid intercepting named scripts (though I'm not 100% sure this env var is documented — it may be an artifact of our specific setup)
3. **Replacing the shell**: `exec xonsh` swaps out the inherited shell for our preferred one

### Gotchas with this approach

For named scripts that need bash-specific helpers, we still have to work around it:

```json
{
  "shell": {
    "scripts": {
      "playwright-cli": ". /home/node/.openclaw/devbox-env/lib/openclaw-shellenv.sh && . /home/node/.openclaw/devbox-env/lib/playwright-env.sh && exec /home/node/.openclaw/npm-global/bin/playwright-cli \"\"",
      "cc": ". /home/node/.openclaw/devbox-env/lib/openclaw-shellenv.sh && bash /home/node/.openclaw/workspace/scripts/cc-dispatch.sh \"\""
    }
  }
}
```

Each script has to:
- Source a shared shellenv file (sh-compatible)
- Explicitly invoke `bash` or the required shell
- Pass arguments through with `""`

### Feature request

A native way to specify the shell for both `devbox shell` and `devbox run` would eliminate this boilerplate. The proposed `.shell.exec.shell` and `.shell.exec.run` split would work well for our use case.

Even simpler would be something like:

```json
{
  "shell": {
    "bin": "/nix/store/...-xonsh/bin/xonsh",
    "init_hook": [...]
  }
}
```

This would:
- Use `xonsh` for interactive `devbox shell`
- Use `xonsh -c` for `devbox run <script>`
- Allow bash/other shells as packages with their rcfiles loaded

Thanks for considering this! The `init_hook` + `exec` hack works but it's fragile and interferes with things like shell history (the parent shell never completes its init).
